README.md
# CheckBeat
A simple checkbox-based drum machine.


Changes.md
CheckBeat

Changes For CheckBeat Version 3:

    Now uses XAudio2 for sound playback.  Versions before 3 used DirectMusic, which Microsoft no longer supports.
    Interface now has the rounded button styles.
    Added a help button.
    Controls that don't affect the beat while it's playing are grayed out during playback.
    Includes 18 drum samples instead of the basic 4.

Changes For CheckBeat Version 2:

    Now allows the selection of division length (1/4 to 1/32 note).
    Multithreaded to allow interaction with dialog without disrupting currently playing beat.


resource.h
//{{NO_DEPENDENCIES}}
// Microsoft Visual C++ generated include file.
// Used by DirectNoise.rc
//
#define IDC_MYICON                      2
#define IDD_DIRECTNOISE_DIALOG          102
#define IDD_DRUMMACHINE                 102
#define IDD_DRUMPANEL                   102
#define IDS_APP_TITLE                   103
#define IDD_ABOUTBOX                    103
#define IDM_ABOUT                       104
#define IDM_EXIT                        105
#define IDI_SMALL                       108
#define IDC_DIRECTNOISE                 109
#define IDR_MAINFRAME                   128
#define IDEXIT                          1000
#define IDC_KICK                        1001
#define IDC_C_HIHAT                     1002
#define IDC_KICK_BROWSE                 1003
#define IDC_Snare                       1004
#define IDC_SNARE_BROWSE                1005
#define IDC_HIHAT                       1006
#define IDC_KICK_FILENAME               1007
#define IDC_C_HIHAT_BROWSE              1008
#define IDC_HIHAT_BROWSE                1009
#define IDC_SNARE_FILENAME              1010
#define IDC_C_HIHAT_FILENAME            1011
#define IDC_HIHAT_FILENAME              1012
#define IDC_CHECK1                      1013
#define IDC_CHECK2                      1014
#define IDC_CHECK3                      1015
#define IDC_CHECK4                      1016
#define IDC_CHECK5                      1017
#define IDC_CHECK6                      1018
#define IDC_CHECK7                      1019
#define IDC_CHECK8                      1020
#define IDC_CHECK9                      1021
#define IDC_CHECK10                     1022
#define IDC_CHECK11                     1023
#define IDC_CHECK12                     1024
#define IDC_CHECK13                     1025
#define IDC_CHECK14                     1026
#define IDC_CHECK15                     1027
#define IDC_CHECK16                     1028
#define IDC_CHECK17                     1029
#define IDC_CHECK18                     1030
#define IDC_CHECK19                     1031
#define IDC_CHECK20                     1032
#define IDC_CHECK21                     1033
#define IDC_CHECK22                     1034
#define IDC_CHECK23                     1035
#define IDC_CHECK24                     1036
#define IDC_CHECK25                     1037
#define IDC_CHECK26                     1038
#define IDC_CHECK27                     1039
#define IDC_CHECK28                     1040
#define IDC_CHECK29                     1041
#define IDC_CHECK30                     1042
#define IDC_CHECK31                     1043
#define IDC_CHECK32                     1044
#define IDC_CHECK33                     1045
#define IDC_CHECK34                     1046
#define IDC_CHECK35                     1047
#define IDC_CHECK36                     1048
#define IDC_CHECK37                     1049
#define IDC_CHECK38                     1050
#define IDC_CHECK39                     1051
#define IDC_CHECK40                     1052
#define IDC_CHECK41                     1053
#define IDC_CHECK42                     1054
#define IDC_CHECK43                     1055
#define IDC_CHECK44                     1056
#define IDC_CHECK45                     1057
#define IDC_CHECK46                     1058
#define IDC_CHECK47                     1059
#define IDC_CHECK48                     1060
#define IDC_CHECK49                     1061
#define IDC_CHECK50                     1062
#define IDC_CHECK51                     1063
#define IDC_CHECK52                     1064
#define IDC_CHECK53                     1065
#define IDC_CHECK54                     1066
#define IDC_CHECK55                     1067
#define IDC_CHECK56                     1068
#define IDC_CHECK57                     1069
#define IDC_CHECK58                     1070
#define IDC_CHECK59                     1071
#define IDC_CHECK60                     1072
#define IDC_CHECK61                     1073
#define IDC_CHECK62                     1074
#define IDC_CHECK63                     1075
#define IDC_CHECK64                     1076
#define IDC_BPM                         1077
#define IDC_START                       1078
#define IDC_STOP                        1079
#define IDC_BEATS                       1080
#define IDC_COMBO1                      1081
#define IDC_RESOLUTION                  1081
#define IDC_STATIC                      -1

// Next default values for new objects
// 
#ifdef APSTUDIO_INVOKED
#ifndef APSTUDIO_READONLY_SYMBOLS
#define _APS_NO_MFC                     1
#define _APS_NEXT_RESOURCE_VALUE        130
#define _APS_NEXT_COMMAND_VALUE         32771
#define _APS_NEXT_CONTROL_VALUE         1082
#define _APS_NEXT_SYMED_VALUE           110
#endif
#endif


sampledata.h
#ifndef _SAMPLEDATA_H_
#define _SAMPLEDATA_H_

#include "wavefile.h"
#include <string>
#ifdef WIN32
#include <xaudio2.h>
#endif

// Contains all information related to a sample, including file information, what key it is mapped to,
// and any information used for playback.

class SampleData
{
public:
    SampleData();
    ~SampleData();
    int _keyboardNote; // Note which sample is mapped to on keyboard
    int _sampleNote; // Pitch of original sample.
    int _volume; // DirectSound volume (attenuation, 0 is max volume).
	bool _looping; // Does the sample repeat after playing through?
	bool _userSpecified; // Has this been specified by the user or is it autogenerated?
    std::string _filename; // File containing sample.
	WaveFile* _waveFile; // Actual wave file data for sample.  Should be a pointer
	                    // with reference counting in order to save memory.  That
	                    // comes later.
#ifdef WIN32
	IXAudio2SourceVoice* _sourceVoice; // The actual playback data.
	XAUDIO2_BUFFER* _buffer;
	void CreateSource(IXAudio2* xAudio2); // Creates an IXAudio2SourceVoice from the wave file.
	static std::string GetXAudioErrorString(HRESULT hr); // Gets an error string for an XAudio HRESULT.
	void DestroySource(); // Destroys source and clears buffer.
#endif
};

#endif


DirectNoise.h
#pragma once

#include "resource.h"
#ifdef WIN32
#include "stdafx.h"
#include "Commdlg.h"
#define INITGUID
//#include <dmusici.h>
#endif
#define MAX_LOADSTRING 100

class CheckBeat
{
public:
	CheckBeat();
	~CheckBeat();
	void Run();
	DWORD WINAPI Run( void* arg );
	IDirectMusicLoader8*		_pLoader;
	IDirectMusicPerformance8*	_pPerformance;
	IDirectMusicSegment8*		_pSegment1;
	IDirectMusicSegment8*		_pSegment2;
	IDirectMusicSegment8*		_pSegment3;
	IDirectMusicSegment8*		_pSegment4;
	bool done;
	bool playing;
	int step; // Keep track of which sequencer step we are on.
	int steps; // Keep track of total steps
	int bpm;
	int division; // Parts of a quarter note per beat. 1 = 1/4, 2 = 1/8, 4 = 1/16, 8 = 1/32.
	LARGE_INTEGER currtime;
	LARGE_INTEGER lasttime;
	LARGE_INTEGER tickspersec;
	HWND hWnd;
};


sampledata.cpp
#include "sampledata.h"

/**
* Constructor.
*/
SampleData::SampleData()
{
	_waveFile = NULL;
    _keyboardNote = 64;
    _sampleNote = 64;
    _volume = 127;
	_userSpecified = false;
	_looping = false;
	_sourceVoice = NULL;
	_buffer = NULL;
}

/**
* Destructor.  Cleans up pointers.
*/
SampleData::~SampleData()
{
	DestroySource();
	if( _waveFile != NULL )
	{
		delete _waveFile;
	}
}

/**
* Creates an XAudio2 source from the wavefile data.
*/
void SampleData::CreateSource(IXAudio2* xAudio2)
{
	if( _waveFile == NULL )
		return;
	if( _buffer != NULL || _sourceVoice != NULL )
	{
		DestroySource();
	}
	HRESULT hr = xAudio2->CreateSourceVoice( &_sourceVoice, _waveFile->GetWaveFormatEx(), XAUDIO2_VOICE_USEFILTER, 64.0, NULL, NULL, NULL );
	if( FAILED( hr ))
	{
		MessageBox(NULL, TEXT("Unable to create source voice: %d (%x) [%s]"), TEXT("Create Source Error"), MB_OK);
	}
	_buffer = _waveFile->GetXAudio2Buffer();
	if( _looping )
		_buffer->LoopCount = XAUDIO2_LOOP_INFINITE;
}

/**
* Destroys the source associated with this sample and clears the audio data buffer.
*/
void SampleData::DestroySource()
{
	if( _buffer != NULL )
	{
		if( _buffer->pAudioData != NULL )
			delete[] _buffer->pAudioData;
		delete _buffer;
		_buffer = NULL;
	}
	if( _sourceVoice != NULL )
	{
		_sourceVoice->DestroyVoice();
		_sourceVoice = NULL;
	}
}

/**
* Gets the XAudio error string from an HRESULT.
*/
std::string SampleData::GetXAudioErrorString(HRESULT hr)
{
	switch( hr )
	{
		case XAUDIO2_E_INVALID_CALL:
			return "Invalid XAudio2 API call.";
			break;
		case XAUDIO2_E_XMA_DECODER_ERROR:
			return "The Xbox 360 XMA hardware suffered an unrecoverable error.";
			break;
		case XAUDIO2_E_XAPO_CREATION_FAILED:
			return "An effect failed to instantiate.";
			break;
		case XAUDIO2_E_DEVICE_INVALIDATED:
			return "An audio device became unusable through being unplugged or some other event.";
			break;
		default:
			return "";
	}
}

CheckBeat.h
#pragma once

#include "resource.h"
#ifdef WIN32
#include "stdafx.h"
#include "Commdlg.h"
#include <XAudio2.h>
#endif
#define INITGUID
#include "sampledata.h"
#define MAX_LOADSTRING 100

#define NUM_DRUMS 4

class CheckBeat
{
public:
	CheckBeat();
	~CheckBeat();
	void Run();
	bool SetupAudio();
	void PlayNote(int noteNumber);
	void StopNote(int noteNumber);
#ifdef WIN32
	DWORD WINAPI Run( void* arg );
	IXAudio2* _xAudio2;
	IXAudio2MasteringVoice*	_masteringVoice;
	IXAudio2SourceVoice* _sourceVoice[4];
	HWND hWnd;
	LARGE_INTEGER currtime;
	LARGE_INTEGER lasttime;
	LARGE_INTEGER tickspersec;
#endif
	//WaveFile* _sample[NUM_DRUMS];
	SampleData* _sample[NUM_DRUMS];
	bool _playing[NUM_DRUMS];
	bool done;
	bool playing;
	int step; // Keep track of which sequencer step we are on.
	int steps; // Keep track of total steps
	int bpm;
	int division; // Parts of a quarter note per beat. 1 = 1/4, 2 = 1/8, 4 = 1/16, 8 = 1/32.
};
